### PR TITLE
Generate gapic yaml config for Discovery docs

### PR DESCRIPTION
**Pull request by andreamlin**
Includes a new baseline test for the generated gapic config.
The only differences between the generated config and the one actually used in discogapic gen are:
- the generated config is littered with TODO comments, the ones that remind the user to manually configure settings if needed
- the language settings, which must be (trivially) hand-edited anyway

### CODE REVIEW

**Comment by andreamlin** [comment]
PTAL.

File: src/main/java/com/google/api/codegen/DocumentGenerator.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/** Generates an in-memory Document model object from a Discovery document filepath. */
+public class DocumentGenerator {
+  private static Document createDocument(String discoveryDocPath) throws IOException {
+    if (!new File(discoveryDocPath).exists()) {
+      throw new FileNotFoundException();
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/DocumentGenerator.java]: The exception would be more helpful if you provided the path that was not found.

File: src/main/java/com/google/api/codegen/DocumentGenerator.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    Reader reader = new InputStreamReader(new FileInputStream(new File(discoveryDocPath)));
+    ObjectMapper mapper = new ObjectMapper();
+    JsonNode root = mapper.readTree(reader);
+    return Document.from(new DiscoveryNode(root));
+  }
+
+  public static Document createDocumentAndLog(String discoveryDocPath, DiagCollector diagCollector)
+      throws IOException {
+    // Prevent INFO messages from polluting the log.
+    Logger.getLogger("").setLevel(Level.WARNING);
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/DocumentGenerator.java]: This should probably be done at a higher level (close to where the tool or app is started)

File: src/main/java/com/google/api/codegen/DocumentGenerator.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+    Document document = null;
+    try {
+      document = DocumentGenerator.createDocument(discoveryDocPath);
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/DocumentGenerator.java]: I don't think you need the class qualifier here since you're already in a static method.

File: src/main/java/com/google/api/codegen/config/DiscoveryMethodModel.java [hunk 4]
```diff
@@ -229,4 +258,37 @@ public boolean hasReturnValue() {
...
+    }
+
+    List<FieldModel> outputField;
+    if (method.response() != null && !Strings.isNullOrEmpty(method.response().reference())) {
+      FieldModel fieldModel = new DiscoveryField(method.response().dereference(), null);
+      outputField = Lists.newArrayList(fieldModel);
+    } else {
+      outputField = new ArrayList<>();
+    }
+    outputFields = ImmutableList.copyOf(outputField);
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/config/DiscoveryMethodModel.java]: Why not just make an ImmutableList from the beginning instead of converting?

File: src/main/java/com/google/api/codegen/config/DiscoveryMethodModel.java [hunk 5]
```diff
@@ -229,4 +258,37 @@ public boolean hasReturnValue() {
...
+    return outputFields;
+  }
+
+  @Override
+  /**
+   * Return if this method, as an HTTP method, is idempotent. Based off {@link
+   * com.google.api.tools.framework.aspects.http.model.MethodKind}.
+   */
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/config/DiscoveryMethodModel.java]: The annotation should be after the doc

File: src/main/java/com/google/api/codegen/config/DiscoveryMethodModel.java [hunk 6]
```diff
@@ -229,4 +258,37 @@ public boolean hasReturnValue() {
...
+  public boolean isIdempotent() {
+    Set<String> idempotentHttpMethods = Sets.newHashSet("GET", "HEAD", "PUT", "DELETE");
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/config/DiscoveryMethodModel.java]: Save this set as a static member so it doesn't need to be rebuilt every time

File: src/main/java/com/google/api/codegen/configgen/DiscoConfigGeneratorApi.java [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+  protected void process() throws Exception {
+    String outputPath = options.get(OUTPUT_FILE);
+    Map<String, Doc> outputFiles = generateConfig(outputPath);
+    ToolUtil.writeFiles(outputFiles, "");
+  }
+
+  private Map<String, Doc> generateConfig(String outputPath) {
+    ViewModel viewModel = new DiscoConfigTransformer().generateConfig(document, outputPath);
+    return ImmutableMap.of(
+        outputPath, new CommonSnippetSetRunner(new CommonRenderingUtil()).generate(viewModel));
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/configgen/DiscoConfigGeneratorApi.java]: I think it would be more readable to do the generation on a separate line from the construction of the return value.

File: src/main/java/com/google/api/codegen/configgen/GrpcPagingParameters.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.api.codegen.configgen;
+
+import com.google.common.collect.ImmutableList;
+import java.util.List;
+
+/** Names of paging parameters used in gRPC. */
+public class GrpcPagingParameters implements PagingParameters {
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/configgen/GrpcPagingParameters.java]: Maybe instead `ProtoPagingParameters`?

File: src/main/java/com/google/api/codegen/configgen/GrpcPagingParameters.java [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+    return PARAMETER_MAX_RESULTS;
+  }
+
+  @Override
+  public String getNameForNextPageToken() {
+    return PARAMETER_NEXT_PAGE_TOKEN;
+  }
+
+  @Override
+  public List<String> ignoredParameters() {
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/configgen/GrpcPagingParameters.java]: why not `getIgnoredParameters()`?

File: src/main/java/com/google/api/codegen/configgen/transformer/MethodTransformer.java [hunk 10]
```diff
@@ -142,6 +162,9 @@ private PageStreamingResponseView generatePageStreamingResponse(Method method) {
               method.getSimpleName());
           break;
         }
+      } else if (field.getApiSource().equals(ApiSource.DISCOVERY)) {
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/configgen/transformer/MethodTransformer.java]: Is there any way to base this off of data from pagingParameters instead of making a discovery-specific check?

File: src/main/java/com/google/api/codegen/configgen/transformer/RetryTransformer.java [hunk 11]
```diff
@@ -27,12 +28,23 @@
...
+    switch (transportProtocol) {
+      case HTTP:
+        return generateRetryCodesHttp();
+      case GRPC:
+      default:
+        return generateRetryCodesGrpc();
+    }
+  }
+
+  private List<RetryCodeView> generateRetryCodesGrpc() {
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/configgen/transformer/RetryTransformer.java]: I'd prefer to not have grpc / http specific logic in this general `RetryTransformer` class if it's not too difficult.

File: src/main/java/com/google/api/codegen/configgen/viewmodel/MethodView.java [hunk 12]
```diff
@@ -39,6 +40,15 @@ public boolean hasFlattening() {
   /** Turns on or off generation of a method whose sole parameter is a request object. */
   public abstract boolean requestObjectMethod();
 
+  /** The resource name treatment. */
+  @Nullable
+  public abstract ResourceNameTreatment resourceNameTreatment();
+
+  /** True of the pageStreaming property exists. */
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/configgen/viewmodel/MethodView.java]: s/of/if/

File: src/main/java/com/google/api/codegen/tools/DiscoToolDriverBase.java [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+
+/**
+ * Abstract base class for drivers running tools based on the framework.
+ *
+ * <p>Uses {@link ToolOptions} to pass arguments to the driver.
+ */
+public abstract class DiscoToolDriverBase extends GenericToolDriverBase {
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/tools/DiscoToolDriverBase.java]: I'd prefer not to have this functionality as an extra layer in the inheritance. Could we push the functionality into a utility class?

File: src/test/java/com/google/api/codegen/testdata/discogapic/java/java_simplecompute.v1.json.baseline [hunk 14]
```diff
@@ -3112,11 +3112,9 @@ public final class DeleteAddressHttpRequest implements ApiMessage {
     }
 
     public final AddressName getAddressAsAddressName() {
-      java.lang.String str = getAddress();
-      if (str.isEmpty()) {
-        return null;
-      }
-      return AddressName.parse(str);
+      return AddressName.create(
+          getAddress(),        getProject(),        getRegion()
```
Comment (garrettjonesgoogle) [inline src/test/java/com/google/api/codegen/testdata/discogapic/java/java_simplecompute.v1.json.baseline]: Could you change the template to add a newline on the end of each of these? Formatting doesn't have to be perfect in our generated code since we run the formatter, but I'd like it to be a little better than this.

File: src/test/java/com/google/api/codegen/testdata/simplecompute_gapic.yaml [hunk 15]
```diff
@@ -39,9 +39,9 @@ interfaces:
   # The entity_name is the name to be used as a basis for generated methods and
   # classes.
   collections:
-  - name_pattern: project/{project}/regions/{region}
+  - name_pattern: \{project}/regions/{region}/addresses
```
[the remark this answers is not in the document]
Comment (andreamlin) [reply src/test/java/com/google/api/codegen/testdata/simplecompute_gapic.yaml]: The yaml reader fails to parse the yaml file if the first bracket isn't escaped.

File: src/main/java/com/google/api/codegen/DocumentGenerator.java (hunk 1 above)
Comment (andreamlin) [reply src/main/java/com/google/api/codegen/DocumentGenerator.java]: Done.

File: src/main/java/com/google/api/codegen/DocumentGenerator.java (hunk 2 above)
Comment (andreamlin) [reply src/main/java/com/google/api/codegen/DocumentGenerator.java]: Done.

File: src/main/java/com/google/api/codegen/DocumentGenerator.java (hunk 3 above)
Comment (andreamlin) [reply src/main/java/com/google/api/codegen/DocumentGenerator.java]: Done.

File: src/main/java/com/google/api/codegen/config/DiscoveryMethodModel.java (hunk 4 above)
Comment (andreamlin) [reply src/main/java/com/google/api/codegen/config/DiscoveryMethodModel.java]: Done.

File: src/main/java/com/google/api/codegen/config/DiscoveryMethodModel.java (hunk 5 above)
Comment (andreamlin) [reply src/main/java/com/google/api/codegen/config/DiscoveryMethodModel.java]: What do you mean by "doc"?

File: src/main/java/com/google/api/codegen/configgen/transformer/MethodTransformer.java (hunk 10 above)
Comment (andreamlin) [reply src/main/java/com/google/api/codegen/configgen/transformer/MethodTransformer.java]: Made this MethodTransformer class an abstract class that can do the bulk of the transformations but leaves it up to subclasses to do ApiSource-specific work.

File: src/main/java/com/google/api/codegen/configgen/transformer/RetryTransformer.java (hunk 11 above)
Comment (andreamlin) [reply src/main/java/com/google/api/codegen/configgen/transformer/RetryTransformer.java]: Done.

File: src/main/java/com/google/api/codegen/configgen/viewmodel/MethodView.java (hunk 12 above)
Comment (andreamlin) [reply src/main/java/com/google/api/codegen/configgen/viewmodel/MethodView.java]: Done.

File: src/main/java/com/google/api/codegen/tools/DiscoToolDriverBase.java (hunk 13 above)
Comment (andreamlin) [reply src/main/java/com/google/api/codegen/tools/DiscoToolDriverBase.java]: Done.

File: src/test/java/com/google/api/codegen/testdata/discogapic/java/java_simplecompute.v1.json.baseline (hunk 14 above)
Comment (andreamlin) [reply src/test/java/com/google/api/codegen/testdata/discogapic/java/java_simplecompute.v1.json.baseline]: :)

File: src/main/java/com/google/api/codegen/config/DiscoveryMethodModel.java [hunk 16]
```diff
@@ -26,16 +26,15 @@
...
+import com.google.common.collect.ImmutableList.Builder;
+import com.google.common.collect.ImmutableSet;
 import java.util.LinkedList;
 import java.util.List;
 import java.util.Map;
-import java.util.Set;
 
 /** A wrapper around the model of a Discovery Method. */
 public final class DiscoveryMethodModel implements MethodModel {
+  ImmutableSet<String> IDEMPOTENT_HTTP_METHODS = ImmutableSet.of("GET", "HEAD", "PUT", "DELETE");
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/config/DiscoveryMethodModel.java]: use `private`

File: src/main/java/com/google/api/codegen/config/DiscoveryMethodModel.java [hunk 17]
```diff
@@ -265,26 +264,23 @@ public boolean hasReturnValue() {
       return outputFields;
     }
 
-    List<FieldModel> outputField;
+    ImmutableList.Builder outputField = new Builder();
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/config/DiscoveryMethodModel.java]: Don't you want to parameterize this with `<FieldModel>`? (You should be getting a warning in your IDE)

File: src/main/java/com/google/api/codegen/configgen/transformer/MethodTransformer.java [hunk 18]
```diff
@@ -48,6 +47,15 @@ public MethodTransformer(PagingParameters pagingParameters) {
...
   }
 
+  /** Get the ResourceNameTreatment for a method. */
+  @Nullable
+  abstract ResourceNameTreatment getResourceNameTreatment(MethodModel methodModel);
+
+  /** Make the page streaming response view for a method. */
+  @Nullable
+  abstract PageStreamingResponseView generatePageStreamingResponse(
+      PagingParameters pagingParameters, MethodModel method);
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/configgen/transformer/MethodTransformer.java]: I strongly prefer not to use multi-level inheritance with implementations at multiple levels; it makes code flow hard to follow.

File: src/main/java/com/google/api/codegen/config/DiscoveryMethodModel.java (hunk 16 above)
Comment (andreamlin) [reply src/main/java/com/google/api/codegen/config/DiscoveryMethodModel.java]: Done.

File: src/main/java/com/google/api/codegen/config/DiscoveryMethodModel.java (hunk 17 above)
Comment (andreamlin) [reply src/main/java/com/google/api/codegen/config/DiscoveryMethodModel.java]: Done.

File: src/main/java/com/google/api/codegen/configgen/transformer/MethodTransformer.java (hunk 18 above)
Comment (andreamlin) [reply src/main/java/com/google/api/codegen/configgen/transformer/MethodTransformer.java]: Moved the pagingParameters to the subclasses, so this abstract base class doesn't have it as a constructor parameter.

**Comment by andreamlin** [comment]
Addressed comments. PTAL.

**Comment by andreamlin** [comment]
Fixed the inheritance issue in MethodTransformer. PTAL

**Comment by andreamlin** [comment]
Renamed MethodHelperTransformer to InputSpecificMethodTransformer.
PTAL.

**Comment by garrettjonesgoogle** [comment]
LGTM
